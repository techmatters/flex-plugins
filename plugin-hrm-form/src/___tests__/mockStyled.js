/*
 * File used to populate test's scope with mocked styled components
 * Should be imported before the components making use of the styled
 */

jest.mock('../styles/HrmStyles', () => ({
  Box: 'Box',
  Flex: 'Flex',
  Absolute: 'Absolute',
  TabbedFormsContainer: 'TabbedFormsContainer',
  Container: 'Container',
  ErrorText: 'ErrorText',
  CategoryTitle: 'CategoryTitle',
  CategorySubtitleSection: 'CategorySubtitleSection',
  CategoryRequiredText: 'CategoryRequiredText',
  StyledInput: 'StyledInput',
  TextField: 'TextField',
  StyledLabel: 'StyledLabel',
  StyledSelect: 'StyledSelect',
  StyledMenuItem: 'StyledMenuItem',
  StyledNextStepButton: 'StyledNextStepButton',
  CheckboxField: 'CheckboxField',
  StyledCheckboxLabel: 'StyledCheckboxLabel',
  StyledCategoryCheckboxLabel: 'StyledCategoryCheckboxLabel',
  BottomButtonBar: 'BottomButtonBar',
  NameFields: 'NameFields',
  ColumnarBlock: 'ColumnarBlock',
  TwoColumnLayout: 'TwoColumnLayout',
  CategoryCheckboxField: 'CategoryCheckboxField',
  StyledCategoryCheckbox: 'StyledCategoryCheckbox',
  ToggleViewButton: 'ToggleViewButton',
  CategoriesWrapper: 'CategoriesWrapper',
  SubcategoriesWrapper: 'SubcategoriesWrapper',
  StyledTabs: 'StyledTabs',
  StyledTab: 'StyledTab',
  StyledSearchTab: 'StyledSearchTab',
  Row: 'Row',
  FontOpenSans: 'FontOpenSans',
  HiddenText: 'HiddenText',
  TransferStyledButton: 'TransferStyledButton',
  HeaderContainer: 'HeaderContainer',
  StyledIcon: () => 'StyledIcon',
  addHover: () => 'addHover',
  PaginationRow: 'PaginationRow',
  AddTaskIconContainer: 'AddTaskIconContainer',
  AddTaskIcon: 'AddTaskIcon',
  AddTaskContent: 'AddTaskContent',
  AddTaskText: 'AddTaskText',
  AddTaskButtonBase: 'AddTaskButtonBase',
  FormItem: 'FormItem',
  FormLabel: 'FormLabel',
  DependentSelectLabel: 'DependentSelectLabel',
  FormError: 'FormError',
  FormInput: 'FormInput',
  FormDateInput: 'FormDateInput',
  FormTimeInput: 'FormTimeInput',
  FormTextArea: 'FormTextArea',
  FormCheckBoxWrapper: 'FormCheckBoxWrapper',
  FormCheckbox: 'FormCheckbox',
  FormMixedCheckbox: 'FormMixedCheckbox',
  FormSelectWrapper: 'FormSelectWrapper',
  FormSelect: 'FormSelect',
  FormOption: 'FormOption',
  CategoryCheckbox: 'CategoryCheckbox',
  CategoryCheckboxLabel: 'CategoryCheckboxLabel',
  CategoryCheckboxField: 'CategoryCheckboxField',
  TaskCanvasOverride: 'TaskCanvasOverride',
  PopoverText: 'PopoverText',
  CannedResponsesContainer: 'CannedResponsesContainer',
  Bold: 'Bold',
  UploadFileLabel: 'UploadFileLabel',
  UploadFileFileName: 'UploadFileFileName',
}));

jest.mock('../styles/search', () => ({
  ConfirmContainer: 'ConfirmContainer',
  BackIcon: 'BackIcon',
  ContactWrapper: 'ContactWrapper',
  ConnectIcon: 'ConnectIcon',
  ContactButtonsWrapper: 'ContactButtonsWrapper',
  StyledLink: 'StyledLink',
  ContactTag: 'ContactTag',
  ConfirmText: 'ConfirmText',
  CancelButton: 'CancelButton',
  SilentText: 'SilentText',
  PrevNameText: 'PrevNameText',
  SummaryText: 'SummaryText',
  ShortSummaryText: 'ShortSummaryText',
  CounselorText: 'CounselorText',
  CaseFooter: 'CaseFooter',
  CaseFooterText: 'CaseFooterText',
  DateText: 'DateText',
  TagsWrapper: 'TagsWrapper',
  TagText: 'TagText',
  TagMiddleDot: 'TagMiddleDot',
  ContactDetailsIcon: () => 'ContactDetailsIcon',
  DetailsContainer: 'DetailsContainer',
  SectionTitleContainer: 'SectionTitleContainer',
  NameContainer: 'NameContainer',
  BackText: 'BackText',
  DetNameText: 'DetNameText',
  SectionTitleText: 'SectionTitleText',
  SectionDescriptionText: 'SectionDescriptionText',
  SectionValueText: 'SectionValueText',
  ResultsHeader: 'ResultsHeader',
  ListContainer: 'ListContainer',
  ScrollableList: 'ScrollableList',
  StyledButtonBase: 'StyledButtonBase',
  StyledFormControlLabel: 'StyledFormControlLabel',
  StyledSwitch: 'StyledSwitch',
  SwitchLabel: 'SwitchLabel',
  StyledLink: 'StyledLink',
  StyledTabs: 'StyledTabs',
  StyledResultsContainer: 'StyledResultsContainer',
  StyledResultsText: 'StyledResultsText',
  StyledTabLabel: 'StyledTabLabel',
  StyledFolderIcon: 'StyledFolderIcon',
  BoldText: 'BoldText',
  SearchResults: 'SearchResults',
  CaseHeaderContainer: 'CaseHeaderContainer',
  CaseHeaderCaseId: 'CaseHeaderCaseId',
  CaseHeaderChildName: 'CaseHeaderChildName',
  CaseSummaryContainer: 'CaseSummaryContainer',
  CaseWrapper: 'CaseWrapper',
  SearchTitle: 'SearchTitle',
  StandaloneSearchContainer: 'StandaloneSearchContainer',
  StyledCount: 'StyledCount',
  StyledContactResultsHeader: 'StyledContactResultsHeader',
  StyledCaseResultsHeader: 'StyledCaseResultsHeader',
}));

jest.mock('../styles/callTypeButtons', () => ({
  Container: 'Container',
  Label: 'Label',
  DataCallTypeButton: 'DataCallTypeButton',
  NonDataCallTypeButton: 'NonDataCallTypeButton',
  CloseTaskDialog: 'CloseTaskDialog',
  CloseTaskDialogText: 'CloseTaskDialogText',
  ConfirmButton: 'ConfirmButton',
  CancelButton: 'CancelButton',
  CloseButton: 'CloseButton',
  NonDataCallTypeDialogContainer: 'NonDataCallTypeDialogContainer',
}));

jest.mock('../styles/queuesStatus', () => ({
  Container: 'Container',
  HeaderContainer: 'HeaderContainer',
  QueuesContainer: 'QueuesContainer',
  QueueName: 'QueueName',
  ChannelColumn: 'ChannelColumn',
  ChannelBox: 'ChannelBox',
  ChannelLabel: 'ChannelLabel',
  WaitTimeLabel: 'WaitTimeLabel',
  WaitTimeValue: 'WaitTimeValue',
}));

jest.mock('../styles/case', () => ({
  CaseContainer: 'CaseContainer',
  CaseActionContainer: 'CaseActionContainer',
  CaseActionFormContainer: 'CaseActionFormContainer',
  CenteredContainer: 'CenteredContainer',
  CaseSectionFont: 'CaseSectionFont',
  ViewButton: 'ViewButton',
  CaseAddButtonFont: 'CaseAddButtonFont',
  CaseActionTitle: 'CaseActionTitle',
  CaseActionDetailFont: 'CaseActionDetailFont',
  BaseTextArea: 'BaseTextArea',
  TimelineRow: 'TimelineRow',
  TimelineDate: 'TimelineDate',
  TimelineText: 'TimelineText',
  InformationBoldText: 'InformationBoldText',
  PlaceHolderText: 'PlaceHolderText',
  TimelineIconContainer: 'TimelineIconContainer',
  CaseSummaryTextArea: 'CaseSummaryTextArea',
  RowItemContainer: 'RowItemContainer',
  NoteContainer: 'NoteContainer',
  DetailsContainer: 'DetailsContainer',
  DetailEntryText: 'DetailEntryText',
  DetailDescription: 'DetailDescription',
  DetailsHeaderChildName: 'DetailsHeaderChildName',
  DetailsHeaderCaseContainer: 'DetailsHeaderCaseContainer',
  DetailsHeaderCaseId: 'DetailsHeaderCaseId',
  DetailsHeaderOfficeName: 'DetailsHeaderOfficeName',
  StyledInputField: 'StyledInputField',
  StyledSelectWrapper: 'StyledSelectWrapper',
  StyledSelectField: 'StyledSelectField',
  TimelineFileName: 'TimelineFileName',
}));

jest.mock('../styles/previousContactsBanner', () => ({
  YellowBanner: 'YellowBanner',
}));
