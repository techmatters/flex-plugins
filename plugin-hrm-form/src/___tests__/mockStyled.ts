/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

/*
 * File used to populate test's scope with mocked styled components
 * Should be imported before the components making use of the styled
 */

jest.mock('../styles/', () => ({
  Box: 'Box',
  Flex: 'Flex',
  Absolute: 'Absolute',
  TabbedFormsContainer: 'TabbedFormsContainer',
  Container: 'Container',
  ErrorText: 'ErrorText',
  CategoryTitle: 'CategoryTitle',
  CategorySubtitleSection: 'CategorySubtitleSection',
  CategoryRequiredText: 'CategoryRequiredText',
  StyledInput: 'StyledInput',
  TextField: 'TextField',
  StyledLabel: 'StyledLabel',
  StyledSelect: 'StyledSelect',
  StyledMenuItem: 'StyledMenuItem',
  StyledNextStepButton: 'StyledNextStepButton',
  CheckboxField: 'CheckboxField',
  StyledCheckboxLabel: 'StyledCheckboxLabel',
  StyledCategoryCheckboxLabel: 'StyledCategoryCheckboxLabel',
  BottomButtonBar: 'BottomButtonBar',
  NameFields: 'NameFields',
  ColumnarBlock: 'ColumnarBlock',
  TwoColumnLayout: 'TwoColumnLayout',
  CategoryCheckboxField: 'CategoryCheckboxField',
  StyledCategoryCheckbox: 'StyledCategoryCheckbox',
  ToggleViewButton: 'ToggleViewButton',
  CategoriesWrapper: 'CategoriesWrapper',
  SubcategoriesWrapper: 'SubcategoriesWrapper',
  StyledTabs: 'StyledTabs',
  StyledTab: 'StyledTab',
  StyledSearchTab: 'StyledSearchTab',
  Row: 'Row',
  FontOpenSans: 'FontOpenSans',
  HiddenText: 'HiddenText',
  TransferStyledButton: 'TransferStyledButton',
  HeaderContainer: 'HeaderContainer',
  StyledIcon: () => 'StyledIcon',
  addHover: () => 'addHover',
  PaginationRow: 'PaginationRow',
  AddTaskIconContainer: 'AddTaskIconContainer',
  AddTaskIcon: 'AddTaskIcon',
  AddTaskContent: 'AddTaskContent',
  AddTaskText: 'AddTaskText',
  AddTaskButtonBase: 'AddTaskButtonBase',
  FormItem: 'FormItem',
  FormLabel: 'FormLabel',
  DependentSelectLabel: 'DependentSelectLabel',
  FormError: 'FormError',
  FormInput: 'FormInput',
  FormDateInput: 'FormDateInput',
  FormTimeInput: 'FormTimeInput',
  FormTextArea: 'FormTextArea',
  FormCheckBoxWrapper: 'FormCheckBoxWrapper',
  FormCheckbox: 'FormCheckbox',
  FormMixedCheckbox: 'FormMixedCheckbox',
  FormSelectWrapper: 'FormSelectWrapper',
  FormSelect: 'FormSelect',
  FormOption: 'FormOption',
  CategoryCheckbox: 'CategoryCheckbox',
  CategoryCheckboxLabel: 'CategoryCheckboxLabel',
  TaskCanvasOverride: 'TaskCanvasOverride',
  PopoverText: 'PopoverText',
  CannedResponsesContainer: 'CannedResponsesContainer',
  Bold: 'Bold',
  StyledBackButton: 'StyledBackButton',
  UploadFileLabel: 'UploadFileLabel',
  UploadFileFileName: 'UploadFileFileName',
}));

jest.mock('../components/search/styles', () => ({
  ConfirmContainer: 'ConfirmContainer',
  BackIcon: 'BackIcon',
  ContactWrapper: 'ContactWrapper',
  ConnectIcon: 'ConnectIcon',
  ContactButtonsWrapper: 'ContactButtonsWrapper',
  CategoryChipContainer: 'CategoryChipContainer',
  ConfirmText: 'ConfirmText',
  CancelButton: 'CancelButton',
  SilentText: 'SilentText',
  PrevNameText: 'PrevNameText',
  SummaryText: 'SummaryText',
  ShortSummaryText: 'ShortSummaryText',
  CounselorText: 'CounselorText',
  CaseFooter: 'CaseFooter',
  CaseFooterText: 'CaseFooterText',
  DateText: 'DateText',
  TagsWrapper: 'TagsWrapper',
  ChipText: 'ChipText',
  TagMiddleDot: 'TagMiddleDot',
  ContactDetailsIcon: () => 'ContactDetailsIcon',
  DetailsContainer: 'DetailsContainer',
  SectionTitleContainer: 'SectionTitleContainer',
  SectionTitleButton: 'SectionTitleButton',
  SectionCollapse: 'SectionCollapse',
  BackText: 'BackText',
  DetNameText: 'DetNameText',
  SectionTitleText: 'SectionTitleText',
  SectionDescriptionText: 'SectionDescriptionText',
  SectionValueText: 'SectionValueText',
  ResultsHeader: 'ResultsHeader',
  ListContainer: 'ListContainer',
  ResultsSubheader: 'ResultsSubheader',
  StyledButtonBase: 'StyledButtonBase',
  StyledFormControlLabel: 'StyledFormControlLabel',
  StyledSwitch: 'StyledSwitch',
  SwitchLabel: 'SwitchLabel',
  StyledLink: 'StyledLink',
  StyledTabs: 'StyledTabs',
  StyledResultsContainer: 'StyledResultsContainer',
  StyledResultsText: 'StyledResultsText',
  StyledTabLabel: 'StyledTabLabel',
  StyledFolderIcon: 'StyledFolderIcon',
  BoldText: 'BoldText',
  SearchResults: 'SearchResults',
  CaseHeaderContainer: 'CaseHeaderContainer',
  CaseHeaderCaseId: 'CaseHeaderCaseId',
  CaseHeaderChildName: 'CaseHeaderChildName',
  CaseSummaryContainer: 'CaseSummaryContainer',
  CaseWrapper: 'CaseWrapper',
  SearchTitle: 'SearchTitle',
  StandaloneSearchContainer: 'StandaloneSearchContainer',
  StyledCount: 'StyledCount',
  StyledContactResultsHeader: 'StyledContactResultsHeader',
  StyledCaseResultsHeader: 'StyledCaseResultsHeader',
}));

jest.mock('../components/callTypeButtons/styles', () => ({
  Container: 'Container',
  Label: 'Label',
  DataCallTypeButton: 'DataCallTypeButton',
  NonDataCallTypeButton: 'NonDataCallTypeButton',
  CloseTaskDialog: 'CloseTaskDialog',
  CloseTaskDialogText: 'CloseTaskDialogText',
  ConfirmButton: 'ConfirmButton',
  CancelButton: 'CancelButton',
  CloseButton: 'CloseButton',
  NonDataCallTypeDialogContainer: 'NonDataCallTypeDialogContainer',
}));

jest.mock('../components/queuesStatus/styles', () => ({
  Container: 'Container',
  HeaderContainer: 'HeaderContainer',
  QueuesContainer: 'QueuesContainer',
  QueueName: 'QueueName',
  ChannelColumn: 'ChannelColumn',
  ChannelBox: 'ChannelBox',
  ChannelLabel: 'ChannelLabel',
  WaitTimeLabel: 'WaitTimeLabel',
  WaitTimeValue: 'WaitTimeValue',
}));

jest.mock('../components/case/styles', () => ({
  CaseContainer: 'CaseContainer',
  CaseActionContainer: 'CaseActionContainer',
  CaseActionFormContainer: 'CaseActionFormContainer',
  CaseAddButton: 'CaseAddButton',
  CenteredContainer: 'CenteredContainer',
  CaseSectionFont: 'CaseSectionFont',
  ViewButton: 'ViewButton',
  CaseAddButtonFont: 'CaseAddButtonFont',
  CaseActionTitle: 'CaseActionTitle',
  CaseActionDetailFont: 'CaseActionDetailFont',
  BaseTextArea: 'BaseTextArea',
  TimelineRow: 'TimelineRow',
  TimelineDate: 'TimelineDate',
  TimelineText: 'TimelineText',
  InformationBoldText: 'InformationBoldText',
  PlaceHolderText: 'PlaceHolderText',
  TimelineIconContainer: 'TimelineIconContainer',
  CaseSummaryTextArea: 'CaseSummaryTextArea',
  RowItemContainer: 'RowItemContainer',
  NoteContainer: 'NoteContainer',
  DetailsContainer: 'DetailsContainer',
  DetailEntryText: 'DetailEntryText',
  DetailDescription: 'DetailDescription',
  DetailsHeaderChildName: 'DetailsHeaderChildName',
  DetailsHeaderCaseContainer: 'DetailsHeaderCaseContainer',
  DetailsHeaderCaseId: 'DetailsHeaderCaseId',
  DetailsHeaderOfficeName: 'DetailsHeaderOfficeName',
  StyledInputField: 'StyledInputField',
  StyledSelectWrapper: 'StyledSelectWrapper',
  StyledSelectField: 'StyledSelectField',
  TimelineFileName: 'TimelineFileName',
}));

jest.mock('../components/profile/styles', () => ({
  YellowBannerContainer: 'YellowBannerContainer',
}));

export {};
