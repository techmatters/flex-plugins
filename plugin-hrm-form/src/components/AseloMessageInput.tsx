/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import { MessageInputChildrenProps } from '@twilio/flex-ui-core/src/components/channel/MessageInput/MessageInputImpl';
import React, { Dispatch, useEffect, useState } from 'react';
import { Button, Template, withTheme } from '@twilio/flex-ui';
import { useForm } from 'react-hook-form';
import debounce from 'lodash/debounce';
import { connect } from 'react-redux';

import CannedResponses from './CannedResponses';
import { conversationsBase, namespace, RootState } from '../states';
import {
  MessageSendStatus,
  newSendMessageeAsyncAction,
  newUpdateDraftMessageTextAction,
} from '../states/conversations';
import asyncDispatch from '../states/asyncDispatch';

type MessageProps = Partial<MessageInputChildrenProps>;

const mapDispatchToProps = (
  dispatch: Dispatch<{ type: string } & Record<string, any>>,
  { conversationSid, conversation: { source: conversation } }: MessageProps,
) => ({
  updateDraftMessageText: (text: string) => dispatch(newUpdateDraftMessageTextAction(conversationSid, text)),
  sendMessage: (text: string) => asyncDispatch(dispatch)(newSendMessageeAsyncAction(conversation, text)),
});

const mapStateToProps = (state: RootState, { conversationSid }: MessageProps) => {
  const convoState = state[namespace][conversationsBase][conversationSid];
  return {
    draftText: convoState?.draftMessageText ?? '',
    sendStatus: convoState?.messageSendStatus ?? '',
  };
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type Props = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps> & MessageProps;

const AseloMessageInput: React.FC<Props> = ({
  conversationSid,
  conversation: { source: conversation },
  updateDraftMessageText,
  draftText,
  sendMessage,
  sendStatus,
}) => {
  const { register, handleSubmit, setValue, getValues } = useForm();

  useEffect(() => {
    setValue('messageInputArea', draftText);
    return () => {
      updateDraftMessageText(getValues('messageInputArea'));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [conversationSid]);

  useEffect(() => {
    setValue('messageInputArea', draftText);
    setIsDisabled(sendStatus === MessageSendStatus.SENDING || !draftText);
  }, [draftText, sendStatus, setValue]);

  const [isDisabled, setIsDisabled] = useState(sendStatus === MessageSendStatus.SENDING || !draftText);

  const triggerTyping = debounce(
    () => {
      if (conversation) {
        conversation.typing();
      }
    },
    500,
    {
      leading: true,
      trailing: true,
    },
  );

  const updateSendButtonState = debounce(() => {
    setIsDisabled(sendStatus === MessageSendStatus.SENDING || !getValues('messageInputArea'));
  }, 200);

  const submitMessageForSending = handleSubmit(async () => {
    const message = getValues('messageInputArea');
    /*
     * Deliberately does not use the isDisabled state for this check, because that is set on a debounce.
     * This way the user can hit 'Enter' straight after they finish typing and it will always send.
     */
    if (conversation && message.length && sendStatus !== MessageSendStatus.SENDING) {
      await sendMessage(message);
    }
  });

  const handleChange = () => {
    triggerTyping();
    updateSendButtonState();
  };

  const handleEnterInMessageInput = e => {
    if (e.keyCode === 13 && e.shiftKey === false) {
      e.preventDefault();
      submitMessageForSending();
    }
  };

  return (
    <div key="textarea">
      <textarea
        id="messageInputArea"
        name="messageInputArea"
        ref={ref => {
          register(ref);
        }}
        onChange={handleChange}
        onKeyDown={handleEnterInMessageInput}
      />
      <Button onClick={submitMessageForSending} disabled={isDisabled}>
        <Template code="Send" />
      </Button>
      <CannedResponses key="canned-responses" conversationSid={conversationSid} />
    </div>
  );
};
AseloMessageInput.displayName = 'AseloMessageInput';

export default withTheme(connector(AseloMessageInput));
