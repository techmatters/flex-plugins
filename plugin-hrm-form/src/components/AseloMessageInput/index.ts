export { default as AseloMessageInput } from './AseloMessageInput';
