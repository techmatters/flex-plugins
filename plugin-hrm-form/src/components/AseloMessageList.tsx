import React from 'react';
import type { MessageListChildrenProps } from '@twilio/flex-ui-core/src/components/channel/MessageList/MessageList';
import type { ConversationState } from '@twilio/flex-ui';

import { MessageList } from './Messaging/MessageList';
import type { GroupedMessage } from './Messaging/MessageItem';

type ConversationStateMessage = ConversationState.ConversationState['messages'][number];

const intoGroupedMessage = (m: ConversationStateMessage): GroupedMessage => ({
  body: m.source.body,
  dateCreated: m.source.dateCreated,
  friendlyName: m.authorName,
  from: m.source.author,
  index: m.index,
  isCounselor: m.isFromMe,
  isGroupedWithPrevious: m.groupWithPrevious,
  media: m.bodyAttachment,
  sid: m.source.sid,
  type: m.source.type,
});

const ReplaceMessageList: React.FC<any> = (props: MessageListChildrenProps) => {
  const { messages } = props.conversation;
  const groupedMessages = React.useMemo(() => messages.map(intoGroupedMessage), [messages]);
  return <MessageList messages={groupedMessages} />;
};
ReplaceMessageList.displayName = 'ReplaceMessageList';

export default ReplaceMessageList;
