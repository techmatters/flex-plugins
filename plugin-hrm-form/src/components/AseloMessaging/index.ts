export { AseloMessageInput } from './AseloMessageInput';
export { default as AseloMessageList } from './AseloMessageList';
