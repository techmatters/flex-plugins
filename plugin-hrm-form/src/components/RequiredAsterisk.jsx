/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import React from 'react';

import { fieldType } from '../types';

export const ValidationType = {
  REQUIRED: 'REQUIRED', // Will not be applied if in the callerInformation tab and callType is not caller.  Will not be applied when callType is standalone.
};

const RequiredAsterisk = ({ field }) => {
  const isRequired = field.validation && field.validation.includes(ValidationType.REQUIRED);
  return isRequired && <span style={{ color: 'red' }}>*</span>;
};

RequiredAsterisk.displayName = 'RequiredAsterisk';
RequiredAsterisk.propTypes = {
  field: fieldType.isRequired,
};

export default RequiredAsterisk;
