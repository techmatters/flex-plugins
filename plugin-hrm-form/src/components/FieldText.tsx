/* eslint-disable react/prop-types */
import React from 'react';
import { TextField as TextFields } from '@material-ui/core';

import { StyledInput, StyledLabel, ErrorText, TextField, FormInput, FormLabel } from '../styles/HrmStyles';
import RequiredAsterisk from './RequiredAsterisk';
import { FormFieldType } from './common/forms/types';

type OwnProps = {
  id: string;
  label?: string | JSX.Element;
  placeholder?: string;
  field: FormFieldType;
  rows?: number;
  // Someday we should make this consistent with DefaultEventHandlers in /components/common/forms/types.ts
  handleBlur: React.FocusEventHandler<HTMLDivElement | HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
  handleChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
  handleFocus: React.FocusEventHandler<HTMLDivElement | HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
};

const FieldText: React.FC<OwnProps> = ({
  id,
  label,
  placeholder,
  field,
  rows,
  handleBlur,
  handleChange,
  handleFocus,
  ...rest
}) => (
  <TextField {...rest}>
    <FormLabel htmlFor={id}>
      {label}
      <RequiredAsterisk field={field} />
    </FormLabel>
    <FormInput
      id={id}
      placeholder={placeholder}
      error={field.error !== null}
      value={field.value}
      /*
       * multiline={Boolean(rows)}
       * rows={rows}
       */
      onBlur={handleBlur}
      onChange={handleChange}
      onFocus={handleFocus}
    />
    {field.error && <ErrorText>{field.error}</ErrorText>}
  </TextField>
);

FieldText.displayName = 'FieldText';

FieldText.defaultProps = {
  label: '',
  placeholder: '',
  rows: undefined,
};

export default FieldText;
