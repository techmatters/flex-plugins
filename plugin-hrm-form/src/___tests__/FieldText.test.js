/**
 * Copyright (C) 2021-2023 Technology Matters
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see https://www.gnu.org/licenses/.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import './mockStyled';

import FieldText from '../components/FieldText';
import { ValidationType } from '../components/RequiredAsterisk';

// TODO: Improve assertion for all tests below
test('render basic FieldText', () => {
  const firstName = {
    value: '',
    touched: false,
    error: null,
    validation: null,
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render required FieldText', () => {
  const firstName = {
    value: '',
    touched: false,
    error: null,
    validation: [ValidationType.REQUIRED],
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render FieldText with errors', () => {
  const firstName = {
    value: '',
    touched: true,
    error: 'This field is required',
    validation: [ValidationType.REQUIRED],
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
