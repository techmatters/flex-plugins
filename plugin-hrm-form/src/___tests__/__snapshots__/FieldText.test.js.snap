// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render FieldText with errors 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <FormLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      *
    </span>
  </FormLabel>
  <FormInput
    error={true}
    id="CallerInformation_FirstName"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
  <ErrorText>
    This field is required
  </ErrorText>
</TextField>
`;

exports[`render basic FieldText 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <FormLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
  </FormLabel>
  <FormInput
    error={false}
    id="CallerInformation_FirstName"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
</TextField>
`;

exports[`render required FieldText 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <FormLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      *
    </span>
  </FormLabel>
  <FormInput
    error={false}
    id="CallerInformation_FirstName"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
</TextField>
`;
